### PR TITLE
Split `Finding` in two different classes

### PR DESCRIPTION
**Pull request by BraisGabin**
This is the first part to implement #6854

In this part I just split the classes and clean a bit them. We had a lot of things that we didn't need and hacks to avoid compatibility.

For now, we have the normal `Finding` that is what a `Rule` reports and then `Finding2` that is what detekt reports after collect all the necessary metadata.

`Finding2` is a temporal name! It will be changed to something else in future PRs. Probably `Issue` but for now I will keep it as `Finding2` meanwhile I do the other refactors that this PR allow us.

### CODE REVIEW

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/Analyzer.kt [hunk 1]
```diff
@@ -214,3 +220,22 @@ internal fun ProcessingSpec.workaroundConfiguration(config: Config): Config = wi
...
+
+private fun Finding.toFinding2(): Finding2 {
+    return when (this) {
+        is CorrectableCodeSmell -> Finding2Impl(issue, entity, message, references, severity, autoCorrectEnabled)
+
+        is CodeSmell -> Finding2Impl(issue, entity, message, references, severity)
+
+        else -> error("wtf?")
+    }
+}
```
Comment (BraisGabin) [inline detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/Analyzer.kt:232]: This code is *really* bad. I know. This will change in the next PRs.

**Comment by marschwar** [comment]
To be absolutely honest I don't know if this is such a helpful next step forwards. Would it be possible to have the rule return a `CodeSmell` instead of a `Finding` and model the `CodeSmell` to only hold the necessary information and do the aggregation to a finding in the core?

**Review by marschwar — APPROVED** [review]
I think we have the same general understanding what the whole think should look like once it is finished. It just feels weird to approve a PR that adds throw away code. But the alternative of a giant PR that changes everything at once is even less ideal.

**Comment by BraisGabin** [comment]
> I think we have the same general understanding what the whole think should look like once it is finished. It just feels weird to approve a PR that adds throw away code. But the alternative of a giant PR that changes everything at once is even less ideal.

Lately (probably this last year or so) I found out that I'm way faster refactoring writing code that I know that I'm going to remove in the next 4 commits than try to be smart and avoid that step.

**Pull request merged**